File: WildcardMatching.cpp
// Approach - DP
// Time Complexity - O(m*n)
// Space Complexity - O(m*n), can be optimised to O(length of 's').
// Problems Faced - No!
// It runs on Leetcode!
class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.length();
        int m = p.length();
        if(p == "*")
            return true;
        
        vector<vector<bool>> dp(m+1, vector<bool>(n+1));
        
        dp[0][0] = true;
        
        for(int i = 1; i <= m; i++){
            if(p[i-1] == '*')
                dp[i][0] = dp[i-1][0];
        }
        
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                if(p[i-1] == '*')
                    dp[i][j] = dp[i][j-1] || dp[i-1][j];
                else if(p[i-1] == '?' || p[i-1] == s[j-1])
                    dp[i][j] = dp[i-1][j-1];
            }
        }
        return dp[m][n];
    }
};


// Approach - 2-pointer solution
// Time Complexity - O(sl*log(pl)) - average case scenario.
// Space Complexity - O(1)
// Problems Faced - No!
// It runs on Leetcode!

class Solution {
public:
    bool isMatch(string s, string p) {
        int sL = s.length();
        int pL = p.length();
        int sStar = -1;
        int pStar = -1;
        
        int sP = 0;
        int pP = 0;
        
        while(sP != sL){
            if(s[sP] == p[pP] || p[pP] == '?'){
                sP++;
                pP++;
            }
            else if(p[pP] == '*'){
                pStar = pP;
                sStar = sP;
                pP++;
            }
            else if(pStar == -1)
                return false;
            else{
                pP = pStar + 1;
                sP = sStar + 1;
                sStar = sP;
            }
        }
        
        while(pP != pL){
            if(p[pP] != '*')
                return false;
            pP++;
        }
        
        return true;
    }
};

File: src/WildCardMatching.java
// Time Complexity : O(S * P)
// Space Complexity : O(S * P)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * use Dp. Start iterating from 0th index of s and p. check if both are equal.
 * or if the pattern current index is ?, if so, iterate next index of each
 * string. Else if current index of p is *, then three possibilities are
 * possible. 1. update both s and p indices. 2. Update s index. 3. Update p
 * index. Else return false. when both strings meet end index, return true. Else
 * if s reached end, then check remaining characters of p are *. If not return
 * false. If either of them reach end but not the other, then also false.
 *
 */
class Solution {
	public boolean isMatch(String s, String p) {
		int m = s.length();
		int n = p.length();
		Boolean[][] dp = new Boolean[m + 1][n + 1];
		return helper(s, p, dp, 0, 0);
	}

	public boolean helper(String s, String p, Boolean[][] dp, int m, int n) {
		if (m == s.length() && n == p.length())
			return true;
		if (m == s.length()) {
			for (int i = n; i < p.length(); i++) {
				if (p.charAt(i) != '*')
					return false;
			}
			return true;
		}
		if (m == s.length() || n == p.length())
			return false;

		if (dp[m][n] != null)
			return dp[m][n];

		if (s.charAt(m) == p.charAt(n) || p.charAt(n) == '?')
			return dp[m][n] = helper(s, p, dp, m + 1, n + 1);
		if (p.charAt(n) == '*')
			return dp[m][n] = helper(s, p, dp, m + 1, n + 1) || helper(s, p, dp, m + 1, n)
					|| helper(s, p, dp, m, n + 1);
		return dp[m][n] = false;
	}
}
